sdk/src/main/scala/com/horizen/forge/ForgeMessageBuilder.scala

```diff
@@ -242,6 +253,58 @@ class ForgeMessageBuilder(mainchainSynchronizer: MainchainSynchronizer,
       ommers = Ommer.toOmmer(block) +: ommers
     }
 
+    // Update block size with Ommers
+    val ommersSerializer = new ListSerializer[Ommer](OmmerSerializer)
+    blockSize += ommersSerializer.toBytes(ommers.asJava).length
+
+    // Get all needed MainchainBlockReferences from the MC Node
+    // Extract MainchainReferenceData and collect as much as the block can fit
+    val mainchainBlockReferenceDataToRetrieve: Seq[MainchainHeaderHash] = branchPointInfo.referenceDataToInclude
+
+    if(mainchainBlockReferenceDataToRetrieve.nonEmpty)
+      blockSize += 4 // placeholder for the MainchainReferenceData Seq size value
+
+    val mainchainReferenceData: ArrayBuffer[MainchainBlockReferenceData] = ArrayBuffer()
+    // Collect MainchainRefData considering the actor message processing timeout
+    // Note: We may do a lot of websocket `getMainchainBlockReference` operations that are a bit slow,
+    // because they are processed one by one, so we limit requests in time.
```

Comment:
We may define the new websocket request to get multiple RefData at a time similar to `getMainchainBlockHeaders` to avoid timeout cases, but the response message may be enormous.
To be discussed.